File: market-data/crypto-news/examples/python/news.py
#!/usr/bin/env python3
"""
News API Examples - Python
Free Crypto News API - https://github.com/nirholas/free-crypto-news

Examples for all news-related endpoints.
"""

import requests
import json
from typing import Optional, List
from datetime import datetime, timedelta

BASE_URL = "https://cryptocurrency.cv"


# =============================================================================
# GET /api/news - Main News Feed
# =============================================================================

def get_news(limit: int = 20, category: Optional[str] = None, 
             source: Optional[str] = None) -> dict:
    """
    Fetch aggregated news from all sources.
    
    Args:
        limit: Number of articles (default: 20, max: 100)
        category: Filter by category (bitcoin, ethereum, defi, nft, etc.)
        source: Filter by source (coindesk, cointelegraph, etc.)
    
    Returns:
        List of news articles
    """
    params = {"limit": limit}
    if category:
        params["category"] = category
    if source:
        params["source"] = source
    
    response = requests.get(f"{BASE_URL}/api/news", params=params)
    return response.json()


# Example usage
if __name__ == "__main__":
    # Basic fetch
    news = get_news(limit=5)
    print("=== Latest News ===")
    for article in news.get("articles", news)[:5]:
        print(f"- {article.get('title', 'No title')}")
    
    # Filter by category
    bitcoin_news = get_news(limit=5, category="bitcoin")
    print("\n=== Bitcoin News ===")
    for article in bitcoin_news.get("articles", bitcoin_news)[:5]:
        print(f"- {article.get('title', 'No title')}")


# =============================================================================
# GET /api/news/international - International News
# =============================================================================

def get_international_news(lang: Optional[str] = None, 
                           translate: bool = True) -> dict:
    """
    Fetch news from 75 international sources in 18 languages.
    
    Args:
        lang: Language code (ko, zh, ja, es, de, fr, etc.)
        translate: Auto-translate to English (default: True)
    
    Returns:
        International news articles with optional translation
    """
    params = {"translate": str(translate).lower()}
    if lang:
        params["lang"] = lang
    
    response = requests.get(f"{BASE_URL}/api/news/international", params=params)
    return response.json()


# =============================================================================
# POST /api/news/extract - Extract Article Content
# =============================================================================

def extract_article(url: str) -> dict:
    """
    Extract full article content from any URL.
    
    Args:
        url: Article URL to extract content from
    
    Returns:
        Extracted article with title, content, images, etc.
    """
    response = requests.post(
        f"{BASE_URL}/api/news/extract",
        json={"url": url},
        headers={"Content-Type": "application/json"}
    )
    return response.json()


# =============================================================================
# GET /api/news/categories - Available Categories
# =============================================================================

def get_categories() -> dict:
    """
    Get all available news categories with article counts.
    
    Returns:
        List of categories with metadata
    """
    response = requests.get(f"{BASE_URL}/api/news/categories")
    return response.json()


# =============================================================================
# GET /api/bitcoin - Bitcoin-specific News
# =============================================================================

def get_bitcoin_news(limit: int = 20) -> dict:
    """
    Get Bitcoin-specific news and analysis.
    
    Args:
        limit: Number of articles
    
    Returns:
        Bitcoin news articles
    """
    response = requests.get(f"{BASE_URL}/api/bitcoin", params={"limit": limit})
    return response.json()


# =============================================================================
# GET /api/defi - DeFi News
# =============================================================================

def get_defi_news(limit: int = 20) -> dict:
    """
    Get DeFi protocol news and updates.
    
    Args:
        limit: Number of articles
    
    Returns:
        DeFi news articles
    """
    response = requests.get(f"{BASE_URL}/api/defi", params={"limit": limit})
    return response.json()


# =============================================================================
# GET /api/breaking - Breaking News
# =============================================================================

def get_breaking_news() -> dict:
    """
    Get breaking/urgent news from the last hour.
    
    Returns:
        Breaking news articles
    """
    response = requests.get(f"{BASE_URL}/api/breaking")
    return response.json()


# =============================================================================
# GET /api/search - Search News
# =============================================================================

def search_news(query: str, limit: int = 20, 
                from_date: Optional[str] = None,
                to_date: Optional[str] = None) -> dict:
    """
    Full-text search across all news articles.
    
    Args:
        query: Search query
        limit: Number of results
        from_date: Start date (YYYY-MM-DD)
        to_date: End date (YYYY-MM-DD)
    
    Returns:
        Search results
    """
    params = {"q": query, "limit": limit}
    if from_date:
        params["from"] = from_date
    if to_date:
        params["to"] = to_date
    
    response = requests.get(f"{BASE_URL}/api/search", params=params)
    return response.json()


# =============================================================================
# GET /api/trending - Trending Topics
# =============================================================================

def get_trending(limit: int = 10) -> dict:
    """
    Get trending topics and keywords.
    
    Args:
        limit: Number of trending items
    
    Returns:
        Trending topics with scores
    """
    response = requests.get(f"{BASE_URL}/api/trending", params={"limit": limit})
    return response.json()


# =============================================================================
# GET /api/sources - Available Sources
# =============================================================================

def get_sources() -> dict:
    """
    Get all available news sources with metadata.
    
    Returns:
        List of sources with article counts
    """
    response = requests.get(f"{BASE_URL}/api/sources")
    return response.json()


# =============================================================================
# GET /api/digest - AI Daily Digest
# =============================================================================

def get_digest(date: Optional[str] = None) -> dict:
    """
    Get AI-generated daily news digest.
    
    Args:
        date: Date for digest (YYYY-MM-DD), defaults to today
    
    Returns:
        AI-summarized daily digest
    """
    params = {}
    if date:
        params["date"] = date
    
    response = requests.get(f"{BASE_URL}/api/digest", params=params)
    return response.json()


# =============================================================================
# GET /api/tags - Browse by Tags
# =============================================================================

def get_tags() -> dict:
    """
    Get all available tags/topics.
    
    Returns:
        List of tags with article counts
    """
    response = requests.get(f"{BASE_URL}/api/tags")
    return response.json()


def get_tag_articles(slug: str, limit: int = 20) -> dict:
    """
    Get articles for a specific tag.
    
    Args:
        slug: Tag slug (e.g., "bitcoin", "ethereum")
        limit: Number of articles
    
    Returns:
        Articles for the tag
    """
    response = requests.get(f"{BASE_URL}/api/tags/{slug}", params={"limit": limit})
    return response.json()


# =============================================================================
# COMPLETE EXAMPLES
# =============================================================================

if __name__ == "__main__":
    print("\n" + "="*60)
    print("FREE CRYPTO NEWS API - PYTHON EXAMPLES")
    print("="*60)
    
    # 1. Basic News
    print("\n📰 1. Latest News (5 articles)")
    news = get_news(limit=5)
    for i, article in enumerate(news.get("articles", news)[:5], 1):
        print(f"   {i}. {article.get('title', 'N/A')[:60]}...")
    
    # 2. Category Filter
    print("\n₿ 2. Bitcoin News")
    btc = get_bitcoin_news(limit=3)
    for article in btc.get("articles", btc)[:3]:
        print(f"   - {article.get('title', 'N/A')[:60]}...")
    
    # 3. Search
    print("\n🔍 3. Search 'Ethereum ETF'")
    results = search_news("Ethereum ETF", limit=3)
    for article in results.get("articles", results)[:3]:
        print(f"   - {article.get('title', 'N/A')[:60]}...")
    
    # 4. Trending
    print("\n🔥 4. Trending Topics")
    trending = get_trending(limit=5)
    for topic in trending.get("topics", trending)[:5]:
        if isinstance(topic, dict):
            print(f"   - {topic.get('keyword', topic)}")
        else:
            print(f"   - {topic}")
    
    # 5. International
    print("\n🌏 5. Korean News (translated)")
    intl = get_international_news(lang="ko")
    for article in intl.get("articles", intl)[:3]:
        print(f"   - {article.get('title', 'N/A')[:60]}...")
    
    # 6. Breaking
    print("\n⚡ 6. Breaking News")
    breaking = get_breaking_news()
    for article in breaking.get("articles", breaking)[:3]:
        print(f"   - {article.get('title', 'N/A')[:60]}...")
    
    # 7. Digest
    print("\n📋 7. Daily Digest")
    digest = get_digest()
    if isinstance(digest, dict) and "summary" in digest:
        print(f"   {digest['summary'][:200]}...")
    else:
        print("   Digest available via API")
    
    print("\n" + "="*60)
    print("All examples completed! See source code for more options.")
    print("="*60)

File: market-data/crypto-news/examples/javascript/news.js
/**
 * News API Examples - JavaScript/Node.js
 * Free Crypto News API - https://github.com/nirholas/free-crypto-news
 * 
 * Examples for all news-related endpoints.
 */

const BASE_URL = 'https://cryptocurrency.cv';

// =============================================================================
// GET /api/news - Main News Feed
// =============================================================================

/**
 * Fetch aggregated news from all sources.
 * @param {Object} options - Query options
 * @param {number} [options.limit=20] - Number of articles (max: 100)
 * @param {string} [options.category] - Filter by category
 * @param {string} [options.source] - Filter by source
 * @returns {Promise<Object>} News articles
 */
async function getNews({ limit = 20, category, source } = {}) {
  const params = new URLSearchParams({ limit });
  if (category) params.append('category', category);
  if (source) params.append('source', source);
  
  const response = await fetch(`${BASE_URL}/api/news?${params}`);
  return response.json();
}

// =============================================================================
// GET /api/news/international - International News
// =============================================================================

/**
 * Fetch news from 75 international sources in 18 languages.
 * @param {Object} options - Query options
 * @param {string} [options.lang] - Language code (ko, zh, ja, es, de, fr, etc.)
 * @param {boolean} [options.translate=true] - Auto-translate to English
 * @returns {Promise<Object>} International news articles
 */
async function getInternationalNews({ lang, translate = true } = {}) {
  const params = new URLSearchParams({ translate });
  if (lang) params.append('lang', lang);
  
  const response = await fetch(`${BASE_URL}/api/news/international?${params}`);
  return response.json();
}

// =============================================================================
// POST /api/news/extract - Extract Article Content
// =============================================================================

/**
 * Extract full article content from any URL.
 * @param {string} url - Article URL to extract content from
 * @returns {Promise<Object>} Extracted article
 */
async function extractArticle(url) {
  const response = await fetch(`${BASE_URL}/api/news/extract`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ url })
  });
  return response.json();
}

// =============================================================================
// GET /api/news/categories - Available Categories
// =============================================================================

/**
 * Get all available news categories with article counts.
 * @returns {Promise<Object>} Categories list
 */
async function getCategories() {
  const response = await fetch(`${BASE_URL}/api/news/categories`);
  return response.json();
}

// =============================================================================
// GET /api/bitcoin - Bitcoin-specific News
// =============================================================================

/**
 * Get Bitcoin-specific news and analysis.
 * @param {number} [limit=20] - Number of articles
 * @returns {Promise<Object>} Bitcoin news
 */
async function getBitcoinNews(limit = 20) {
  const response = await fetch(`${BASE_URL}/api/bitcoin?limit=${limit}`);
  return response.json();
}

// =============================================================================
// GET /api/defi - DeFi News
// =============================================================================

/**
 * Get DeFi protocol news and updates.
 * @param {number} [limit=20] - Number of articles
 * @returns {Promise<Object>} DeFi news
 */
async function getDefiNews(limit = 20) {
  const response = await fetch(`${BASE_URL}/api/defi?limit=${limit}`);
  return response.json();
}

// =============================================================================
// GET /api/breaking - Breaking News
// =============================================================================

/**
 * Get breaking/urgent news from the last hour.
 * @returns {Promise<Object>} Breaking news
 */
async function getBreakingNews() {
  const response = await fetch(`${BASE_URL}/api/breaking`);
  return response.json();
}

// =============================================================================
// GET /api/search - Search News
// =============================================================================

/**
 * Full-text search across all news articles.
 * @param {string} query - Search query
 * @param {Object} options - Search options
 * @param {number} [options.limit=20] - Number of results
 * @param {string} [options.from] - Start date (YYYY-MM-DD)
 * @param {string} [options.to] - End date (YYYY-MM-DD)
 * @returns {Promise<Object>} Search results
 */
async function searchNews(query, { limit = 20, from, to } = {}) {
  const params = new URLSearchParams({ q: query, limit });
  if (from) params.append('from', from);
  if (to) params.append('to', to);
  
  const response = await fetch(`${BASE_URL}/api/search?${params}`);
  return response.json();
}

// =============================================================================
// GET /api/trending - Trending Topics
// =============================================================================

/**
 * Get trending topics and keywords.
 * @param {number} [limit=10] - Number of trending items
 * @returns {Promise<Object>} Trending topics
 */
async function getTrending(limit = 10) {
  const response = await fetch(`${BASE_URL}/api/trending?limit=${limit}`);
  return response.json();
}

// =============================================================================
// GET /api/sources - Available Sources
// =============================================================================

/**
 * Get all available news sources with metadata.
 * @returns {Promise<Object>} Sources list
 */
async function getSources() {
  const response = await fetch(`${BASE_URL}/api/sources`);
  return response.json();
}

// =============================================================================
// GET /api/digest - AI Daily Digest
// =============================================================================

/**
 * Get AI-generated daily news digest.
 * @param {string} [date] - Date for digest (YYYY-MM-DD)
 * @returns {Promise<Object>} Daily digest
 */
async function getDigest(date) {
  const params = date ? `?date=${date}` : '';
  const response = await fetch(`${BASE_URL}/api/digest${params}`);
  return response.json();
}

// =============================================================================
// GET /api/tags - Browse by Tags
// =============================================================================

/**
 * Get all available tags/topics.
 * @returns {Promise<Object>} Tags list
 */
async function getTags() {
  const response = await fetch(`${BASE_URL}/api/tags`);
  return response.json();
}

/**
 * Get articles for a specific tag.
 * @param {string} slug - Tag slug
 * @param {number} [limit=20] - Number of articles
 * @returns {Promise<Object>} Tagged articles
 */
async function getTagArticles(slug, limit = 20) {
  const response = await fetch(`${BASE_URL}/api/tags/${slug}?limit=${limit}`);
  return response.json();
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
  getNews,
  getInternationalNews,
  extractArticle,
  getCategories,
  getBitcoinNews,
  getDefiNews,
  getBreakingNews,
  searchNews,
  getTrending,
  getSources,
  getDigest,
  getTags,
  getTagArticles,
  BASE_URL
};

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

async function runExamples() {
  console.log('\n' + '='.repeat(60));
  console.log('FREE CRYPTO NEWS API - JAVASCRIPT EXAMPLES');
  console.log('='.repeat(60));

  // 1. Basic News
  console.log('\n📰 1. Latest News (5 articles)');
  const news = await getNews({ limit: 5 });
  const articles = news.articles || news;
  articles.slice(0, 5).forEach((article, i) => {
    console.log(`   ${i + 1}. ${article.title?.substring(0, 60)}...`);
  });

  // 2. Category Filter
  console.log('\n₿ 2. Bitcoin News');
  const btcNews = await getBitcoinNews(3);
  (btcNews.articles || btcNews).slice(0, 3).forEach(article => {
    console.log(`   - ${article.title?.substring(0, 60)}...`);
  });

  // 3. Search
  console.log('\n🔍 3. Search "Ethereum ETF"');
  const searchResults = await searchNews('Ethereum ETF', { limit: 3 });
  (searchResults.articles || searchResults).slice(0, 3).forEach(article => {
    console.log(`   - ${article.title?.substring(0, 60)}...`);
  });

  // 4. Trending
  console.log('\n🔥 4. Trending Topics');
  const trending = await getTrending(5);
  (trending.topics || trending).slice(0, 5).forEach(topic => {
    const keyword = typeof topic === 'object' ? topic.keyword : topic;
    console.log(`   - ${keyword}`);
  });

  // 5. International
  console.log('\n🌏 5. Korean News (translated)');
  const intlNews = await getInternationalNews({ lang: 'ko' });
  (intlNews.articles || intlNews).slice(0, 3).forEach(article => {
    console.log(`   - ${article.title?.substring(0, 60)}...`);
  });

  // 6. Breaking
  console.log('\n⚡ 6. Breaking News');
  const breaking = await getBreakingNews();
  (breaking.articles || breaking).slice(0, 3).forEach(article => {
    console.log(`   - ${article.title?.substring(0, 60)}...`);
  });

  // 7. Sources
  console.log('\n📡 7. Available Sources');
  const sources = await getSources();
  console.log(`   Total sources: ${Array.isArray(sources) ? sources.length : 'See response'}`);

  console.log('\n' + '='.repeat(60));
  console.log('All examples completed!');
  console.log('='.repeat(60));
}

// Run examples if executed directly
if (require.main === module) {
  runExamples().catch(console.error);
}
